public/app/plugins/panel/heatmap-new/utils.ts

```diff
@@ -460,3 +468,64 @@ export const countsToFills = (u: uPlot, seriesIdx: number, palette: string[]) =>
 
   return indexedFills;
 };
+
+export const getHeatmapFields = (dataFrame: DataFrame): Array<Field | undefined> => {
+  const xField: Field | undefined = dataFrame.fields.find((f) => f.name === 'xMin');
+  const yField: Field | undefined = dataFrame.fields.find((f) => f.name === 'yMin');
+  const countField: Field | undefined = dataFrame.fields.find((f) => f.name === 'count');
+
+  return [xField, yField, countField];
+};
+
+export const getHeatmapArrays = (dataFrame: DataFrame): Array<number[] | undefined> => {
+  const [xField, yField, countField] = getHeatmapFields(dataFrame);
+  return [xField?.values.toArray(), yField?.values.toArray(), countField?.values.toArray()];
+};
+
+export const getDataMapping = (
+  heatmapData: HeatmapData,
+  origData: DataFrame,
+  options?: DataMapOptions
+): Array<number[] | null> => {
+  const [fxs, fys, fcounts] = getHeatmapFields(heatmapData.heatmap!);
+  const xos: number[] | undefined = origData.fields.find((f: Field) => f.type === 'time')?.values.toArray();
+  const yos: number[] | undefined = origData.fields.find((f: Field) => f.type === 'number')?.values.toArray();
+
+  if (heatmapData.heatmap?.meta?.type !== DataFrameType.HeatmapScanlines) {
+    return [null];
```

Comment:
this can be first, and should throw an error.  We can catch it and display nicely (but passed with anything else is invalid right now)

for missing xos/yos... I'm not sure the best behavior.  Either error or an array of length `heatmapData.heatmap.length`